### ISSUE TITLE
Discuss: ABI

### ISSUE THREAD

**Issue by pepyakin** [Geal/serverless-wasm#12]
It actually depends on to what degree apps running under `serverless-wasm` should be sandboxed. One one extereme - there is only one pure function that takes input and returns output and on the other — full control of the host machine (Maybe it should be possible to just forward syscalls from wasm to kernel directly).

But the one approach that bugging me is, what if we can use [cloudabi](https://nuxi.nl/cloudabi/) for definition of the host API. This sounds like sorta middle ground between those extremes.

This might be The Most Ambitious Crossover in History...

**Comment by Vengarioth** [comment]
My personal number one use case would be to run untrusted client code.

**Comment by pepyakin** [comment]
I wondering, is this more belongs to https://github.com/CommonWA/cwa-spec

**Comment by Geal** [comment]
for what I imagine, this project will run untrusted code at some point, so good isolation will be important. CloudABI looks interesting :)

**Comment by pepyakin** [comment]
To be clear: As I understand it CloudABI is just the right tool that can allow level good isolation even for the native code! As a bonus we might get some network effect and it's already here and defined.

**Comment by pepyakin** [comment]
1. This is actually the simplest thing. Fork could be implemented by literally cloning memory instance (although it would be more useful for the implementation to employ mmap tricks), table instance and all of the runtime data and launch new thread. The more challenging example would be for example `cloudabi_sys_mem_map()`. Also, usability of `cloudabi_sys_thread_create()` of is quite questionable. It is possible to execute wasm multithreaded, but some semantics aren't defined for now and there are no means of synchronization before threading propolsal accepted. So I thinking more of a subset of it.
2. Correct!
3. Also correct!

**Comment by Vengarioth** [comment]
Thank you for clarifying. Now this comes back to the use case i mentioned first: does the spec have a way to limit the resources a wasm module can allocate? Could i for instance forbid wasm modules to spawn threads(if/when the wasm spec advances to threads) or processes entirely?

Also how does error handling work in this spec?

**Comment by pepyakin** [comment]
>  does the spec have a way to limit the resources a wasm module can allocate

Yes! It allows for imposing additional limits to memory sizes ([link](https://webassembly.github.io/spec/core/appendix/implementation.html#execution)). 

>   Could i for instance forbid wasm modules to spawn threads(if/when the wasm spec advances to threads)

That's a tricky question. In kovan-wasm (ethereum-like blockchain that's capable of running contract in wasm) we imposing some additional limits. For example: we completly ban floats since they can introduce non-determinism (and blockchains require that every program yields absolutly identical results) and this is specifically forbidden by the spec:

> A conforming implementation is not allowed to leave out individual features. However, designated subsets of WebAssembly may be specified in the future.

But I firmly believe that such exclusion should be possible for certain environments, blockchains are the one and IoT devices that might not have FPU is another. (Btw, threads are another feature that won't play nice with blockchains).

The problem is that today is 0 (zero) additional features were delivered so it's not clear how such exclusion mechanisms would work.

> Also how does error handling work in this spec?

Error handling is actually is very opaque. 

For instance, take a look at this [piece](https://webassembly.github.io/spec/core/appendix/embedding.html). 

The is a following note:

> Errors are abstract and unspecific with this definition. Implementations can refine it to carry suitable classifications and diagnostic messages.
